logstash-core/lib/logstash/pipeline_reporter.rb

```diff
@@ -0,0 +1,119 @@
+require 'ostruct'
+
+# encoding: utf-8
+module LogStash; class PipelineReporter
+  attr_reader :logger, :pipeline
+
+  # This is an immutable copy of the pipeline state,
+  # It is a proxy to a hash to allow us to add methods dynamically to the hash
+  class Snapshot
+    def initialize(data)
+      @data = data
+    end
+
+    def to_hash
+      @data
+    end
+
+    def to_simple_hash
+      {"inflight_count" => inflight_count, "stalling_thread_info" => format_threads_by_plugin}
+    end
+
+    def method_missing(meth)
```

Comment:
Not a fan of method_missing lately since it often makes things more difficult to debug. Tell me a story about why Snapshot needs this kind of interface?